### PR TITLE
Allow client node to read `ChainStateView` outside the `ChainWorkerState`

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

The client node currently loads `ChainStateView` directly from storage when it needs to read some information from it. This is inefficient because it requires reloading the `ChainStateView` (which means deserializing its data) every time. It also holds a lock while doing so, preventing others to access the same chain state.

The goal is to keep the `ChainStateView` in memory for a longer amount of time. In order to do so, each `ChainWorkerState` (introduced in #2055) becomes the owner of its respective `ChainStateView`. That means that it will hold a lock on the `ChainStateView` and the client node won't be able to access it as it is currently accessed.

## Proposal

Access the `ChainStateView` through the `WorkerState` in the `LocalNode`. The `ChainStateView` is wrapped inside an `Arc<RwLock<ChainStateView>>`, which allows the `ChainWorkerState` to share a read-only view of the chain state with the client when needed.

## Test Plan

This is an internal refactor, so existing tests should catch any regressions.

## Release Plan

This is an internal refactor, so nothing is needed.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-service/src/node_service.rs [hunk 1]
```diff
@@ -799,20 +801,65 @@ impl ChainStateViewExtension {
...
-struct ChainStateExtendedView<C>(ChainStateViewExtension, Arc<ChainStateView<C>>)
+struct ChainStateExtendedView<C>(ChainStateViewExtension, ReadOnlyChainStateView<C>)
 where
     C: linera_views::common::Context + Clone + Send + Sync + 'static,
     ViewError: From<C::Error>,
     C::Extra: linera_execution::ExecutionRuntimeContext;
 
+/// A wrapper type that allows proxying GraphQL queries to a [`ChainStateView`] that's behind an
+/// [`OwnedRwLockReadGuard`].
+pub struct ReadOnlyChainStateView<C>(OwnedRwLockReadGuard<ChainStateView<C>>)
```
Comment (afck) [inline linera-service/src/node_service.rs:812]: Do we need to change any `async_graphql` cache settings to make sure this is freed as soon as the request is answered?

File: linera-core/src/chain_worker/state.rs [hunk 2]
```diff
@@ -112,9 +129,10 @@ where
         height: BlockHeight,
         index: u32,
     ) -> Result<Option<Event>, WorkerError> {
-        self.ensure_is_active()?;
+        let mut chain = self.chain.write().await;
```
Comment (ma2bd) [inline linera-core/src/chain_worker/state.rs:132]: This should be `self.chain.read()`

File: linera-core/src/chain_worker/state.rs [hunk 3]
```diff
@@ -76,16 +81,27 @@ where
...
+        self.chain_id
+    }
+
+    /// Returns a read-only view of the [`ChainStateView`].
+    ///
+    /// The returned view holds a lock on the chain state, which prevents the worker from changing
+    /// it.
+    pub async fn chain_state_view(
+        &self,
+    ) -> OwnedRwLockReadGuard<ChainStateView<StorageClient::Context>> {
```
Comment (ma2bd) [inline linera-core/src/chain_worker/state.rs:103]: This is arguably a detour as the final state will be an `Arc<ChainStateView<...>>` (if I'm not mistaken)

File: linera-service/src/node_service.rs (hunk 1 above)
Comment (ma2bd) [reply linera-service/src/node_service.rs:812]: See also my other comment

File: linera-core/src/chain_worker/state.rs (hunk 3 above)
Comment (ma2bd) [reply linera-core/src/chain_worker/state.rs:103]: Specifically, the plan was to use `cloen_unchecked`

File: linera-core/src/chain_worker/state.rs (hunk 2 above)
Comment (ma2bd) [reply linera-core/src/chain_worker/state.rs:132]: How about this?
```
let chain = self.chain.read().await();
let mut chain = chain.clone_unchecked();
```

File: linera-service/src/node_service.rs (hunk 1 above)
Comment (jvff) [reply linera-service/src/node_service.rs:812]: I added a `#[graphql(cache_control(no_cache)))]` to `ChainStateView`. Hopefully that's enough :crossed_fingers:

File: linera-core/src/chain_worker/state.rs (hunk 2 above)
Comment (jvff) [reply linera-core/src/chain_worker/state.rs]: Done. I took a deeper look, and there's no need for mutability in these operations :+1: 

I applied the suggested pattern using `clone_unchecked` in the `stage_block_execution` and `query_application` methods, but the problem is that `clone_unchecked` uses a `&mut self` receiver, so a `write` lock is needed anyway, at least to perform the clone. This limits the concurrency available, but also I'm still not sure if it's worth it, because AFAICT it only affects the client, which I think doesn't benefit much from added concurrency here. I can't affirm this with certainty, and we should measure this to decide if we should change it again later.

File: linera-core/src/chain_worker/state.rs (hunk 3 above)
Comment (jvff) [reply linera-core/src/chain_worker/state.rs:103]: I think that to do so we would need to change `clone_unchecked` to receive a `&self`. Otherwise it's worse than the current solution, because it would force getting a write-lock to create the clone, ending up serializing the queries.

File: linera-core/src/chain_worker/state.rs (hunk 2 above)
Comment (ma2bd) [reply linera-core/src/chain_worker/state.rs]: Oh nooo, how badly does clone_unchecked require &mut self ? This def ruins the design that I had in mind all along.

File: linera-core/src/chain_worker/state.rs (hunk 3 above)
Comment (ma2bd) [reply linera-core/src/chain_worker/state.rs:103]: Yes I agree

File: linera-core/src/chain_worker/state.rs (hunk 2 above)
Comment (jvff) [reply linera-core/src/chain_worker/state.rs]: I tried to make it `&self` and ran into four issues:

- https://github.com/linera-io/linera-protocol/blob/bd98cdfb81bd368dd87ae382ca3e0a31954f01b0/linera-views/src/collection_view.rs#L155-L156
- https://github.com/linera-io/linera-protocol/blob/bd98cdfb81bd368dd87ae382ca3e0a31954f01b0/linera-views/src/reentrant_collection_view.rs#L174-L175
- https://github.com/linera-io/linera-protocol/blob/bd98cdfb81bd368dd87ae382ca3e0a31954f01b0/linera-views/src/key_value_store_view.rs#L264
- https://github.com/linera-io/linera-protocol/blob/bd98cdfb81bd368dd87ae382ca3e0a31954f01b0/linera-views/src/hashable_wrapper.rs#L113

In summary, we could make it `&self` if we made the method `async` and accepted the fact that we'd be locking some mutexes in order to clone their contents.

Comment (ma2bd) [reply linera-core/src/chain_worker/state.rs]: We could do that but perhaps let's make sure that we really need it. IIRC our initial plan was for the chain actor to own a (local, mutable) `ChainStateView`.
* This local view would be used by default for everything.
* The first time a read-only view is needed for GraphQL we could clone the local view, wrap it in a `RwLock`, then remember the lock in a local variable of type `Option<Arc<Rwlock<ChainStateView>>>`.
* When the local view is about to be saved, if the rwlock is not `None`, it must be acquired in `write` mode. If the `Arc` counter is 1, then we can reset the lock to `None`. Otherwise (weird case but ok), we will clone the view and update the inner (read-only) view, immediately after saving the local view.

Am I missing something?

Comment (ma2bd) [reply linera-core/src/chain_worker/state.rs]: Mm ok the `Arc` counter may not be `1` if the read-lock is released first, then `Arc` is decremented. But it should about to become `1`.

Comment (jvff) [reply linera-core/src/chain_worker/state.rs]: Maybe I misunderstood the initial plan. I thought that with the `RwLock` we wouldn't need the `clone_unchecked` anymore. But I have now updated the PR with the implementation what you described.

There's no need to check the `Arc` counter, as long as we guarantee that:

- read and write locks are always acquired in `&mut self` methods
- the write lock is only acquired by taking the `Option<Arc<RwLock<_>>>`

As for a better construction, poisoning requires panicking, which feels weird and might be inefficient and barriers require knowing how many times the view will be shared ahead of time. One option is to use a linked-chain of notifiers. It might be cleaner, but likely requires more work to notify-on-drop, and I don't know if it would be more efficient :thinking:

Comment (ma2bd) [reply linera-core/src/chain_worker/state.rs]: Oh right because the `Arc`-holder is hidden in `OwnedRwLockReadGuard`

File: linera-core/src/chain_worker/state.rs [hunk 4]
```diff
@@ -950,6 +973,24 @@ where
             bundle_vecs,
         })
     }
+
+    /// Stores the chain state in persistent storage.
+    ///
+    /// Waits until the [`ChainStateView`] is no longer shared before persisting the changes.
+    async fn save(&mut self) -> Result<(), WorkerError> {
```
Comment (ma2bd) [inline linera-core/src/chain_worker/state.rs:980]: Nice!

**Review by ma2bd — APPROVED** [review]
Epic PR! Thanks!
